### DISCUSSION TITLE
How to map the circuit into LUT and DFF in yosys？

### DISCUSSION THREAD

**Discussion by ghost** [YosysHQ/yosys#3752] [Q&A] [answered]
There is no asynchronous reset port and latch in my circuit, so I want to map it into LUT and DFF only. Now, I find a minimum lib (with INV BUF NAND2 DFF), and my circuit can map to this lib successfully, but I do not know how to further map the INV, BUF, NAND2 into LUT, and therefore I can get a circuit with LUT and DFF only？

**Comment by nakengelhardt** [comment, marked as answer]
You can use e.g. `synth -lut 4` to map the combinatorial logic to 4-LUT. If you only want strictly `$_DFF_[PN]_` (no enable or reset signal on DFF cells) you can use the following script:
```
synth -noabc -top <top_module>
dfflegalize -cell $_DFF_?_ 01
abc -lut 4
opt
```
This should leave you with only `$lut` and `$_DFF_P_/$_DFF_N_` cells in the circuit.

**Reply by ghost** [reply]
Thank you very much! I have tried using synth -lut command, it report the mapping information about this process, and then I tried the write_verilog command, it seems that only the flatten origin netlist is dump to the target file, instead of the connection between the lut and dff? Could you please give me some suggestion about acquiring and dumping out the netlist after mapping? Another problem is that I find when I use synth_ice40, my circuit consume 200 lut4, but when I use abc -lut 4, the same circuit consume about 700 lut4, and I do not known why this happen? Thanks again for your kind help!

**Reply by nakengelhardt** [reply]
By default `write_verilog` will use shift expressions to represent LUTs and processes for DFFs, since these are understood by any tool for post-synthesis simulation without the need to import a cell library. You can use the `-noexpr` option to have it directly instantiate `$lut` and `$_DFF_P_/$_DFF_N_` cells. (The corresponding simulation models are in [simcells.v](https://github.com/YosysHQ/yosys/blob/master/techlibs/common/simcells.v), should you need them.)

As for the difference in LUT count between `synth_ice40` and `synth`, most likely some hard blocks like memory or DSP are inferred by `synth_ice40`, whereas `synth` has to use only FF and LUT for everything. You can check in the `stat` listing at the end of `synth_ice40` for the presence of blocks like `SB_RAM40` or `SB_MAC16`.

**Reply by ghost** [reply]
Thank you very much!

**Reply by Maerd6** [reply]
Hi. I have another quetions about convert xilinx lut structure to logic gates like and, or. Can I manage it using yosys? Can you give me some suggestions?

**Reply by ghost** [reply]
Hi, I think if you can acquire the LUT connection of a circuit, it is easy to convert it to the basic logic gate, for example, the lut4 and it's confgiuration information (16bit) can stand for a truth table, based on the truth table, it is easy to convert it to basic logic gates, but I do not know how to deal this problem in yosys.

**Reply by Maerd6** [reply]
Thank you very much!

**Reply by GPJ-ZFF** [reply]
Hi，I am currently seeking to validate the consistency of the netlist pre and post technology mapping. I believe there is potential to expand the LUT in order to achieve the basic and-OR gate logic function. What would be the next steps? Are there any relevant literature recommendations?
